accord-core/src/main/java/accord/local/Node.java

```diff
@@ -285,31 +286,24 @@ public void onEpochRedundant(Ranges ranges, long epoch)
         topology.onEpochRedundant(ranges, epoch);
     }
 
-    public AsyncChain<?> awaitEpoch(long epoch)
-    {
-        if (topology.hasEpoch(epoch))
-            return AsyncResults.SUCCESS_VOID;
-        configService.fetchTopologyForEpoch(epoch);
-        return topology.awaitEpoch(epoch);
-    }
-
-    public AsyncChain<?> awaitEpoch(@Nullable EpochSupplier epochSupplier)
+    public void withEpoch(EpochSupplier epochSupplier, BiConsumer<Void, Throwable> callback)
     {
         if (epochSupplier == null)
-            return AsyncResults.SUCCESS_VOID;
-        return awaitEpoch(epochSupplier.epoch());
+            callback.accept(null, null);
+        else
+            withEpoch(epochSupplier.epoch(), callback);
     }
 
-    public void withEpoch(long epoch, Runnable runnable)
+    public void withEpoch(long epoch, BiConsumer<Void, Throwable> callback)
     {
         if (topology.hasEpoch(epoch))
         {
-            runnable.run();
+            callback.accept(null, null);
         }
         else
         {
             configService.fetchTopologyForEpoch(epoch);
-            topology.awaitEpoch(epoch).addCallback(runnable).begin(agent);
+            topology.awaitEpoch(epoch).addCallback(callback).begin((ignored1, ignored2) -> {});
```

Comment:
why are you ignoring the error?  If the callback throws we now no longer notify the `agent`?

Reply:
It's redundant because we supply the agent in `TopologyManager` as a listener already. These timeouts produce a lot of not very informative duplicate errors because the exception doesn't actually tell you anything about what witnessed the failure only what failed.

I think I need to do a pass where I make sure the error handlers wrap the exception so they include their stack as well.

Reply:
my point was that if the callback fails the failure is hidden... by keeping the agent we will detect such cases.  Now, because `callback` doesn't finish the chain you would need to do something else... 

thinking about it, why have `callback` and `begin`... why not just use `begin` directly?

Reply:
`begin` is documented as not being safe to call multiple times and the result of `awaitEpoch` might not be a new `AsyncChain`, it might be one that is already started. It happens to be that right now `AsyncResult` doesn't actually error out when `begin` is called.

So by contract this is all pretty problematic and in fact `begin` should not even need to be called. I think the problem is partially that `awaitEpoch` should probably return `AsyncResult` not `AsyncChain` so you don't need to call begin you can just add the callback.

The reason it returns `AsyncChain` is because the `withExecutor` in `TopologyManager.awaitEpoch` converts it to an `AsyncChain`.

It might be better to add an `AsyncResults.withExecutor` and use that?